client/lib/mshots/index.js

```diff
@@ -22,41 +18,49 @@ const querymShotsEndpoint = options => {
 		return;
 	}
 
-	request
-		.get( mShotsEndpointUrl )
-		.responseType( 'blob' )
-		.then( res => {
-			if (
-				( res.type === null || res.type === 'image/gif' ) &&
-				maxRetries > 0 &&
-				currentRetries < maxRetries
-			) {
-				// Still generating the preview or something failed in mShots
-				setTimeout(
-					querymShotsEndpoint.bind( this, {
-						...options,
-						currentRetries: currentRetries + 1,
-					} ),
-					retryTimeout
-				);
-			} else if ( res.type === 'image/jpeg' ) {
-				// Successfully generated the preview
-				try {
-					resolve( window.URL.createObjectURL( res.xhr.response ) );
-				} catch ( e ) {
-					resolve( mShotsEndpointUrl );
-				}
-			} else {
-				reject();
+	try {
+		const response = await fetch( mShotsEndpointUrl, { method: 'GET' } );
+		const contentType = response.ok && response.headers.get( 'Content-Type' );
+
+		if (
+			// This confused me until I tried it out, so I figured I'd leave a comment.
+			// The mshots API takes a while to generate an image if it's a cache miss.
+			// In those situations, it returns a 307 redirecting to a default "loading" image.
+			// That image is a GIF, which is what we're checking for here...
+			( ! contentType || contentType === 'image/gif' ) &&
+			maxRetries > 0 &&
+			currentRetries < maxRetries
+		) {
+			// Still generating the preview or something failed in mShots. Retry.
+			setTimeout(
+				querymShotsEndpoint.bind( this, {
+					...options,
+					currentRetries: currentRetries + 1,
+				} ),
+				retryTimeout
+			);
+		} else if ( contentType === 'image/jpeg' ) {
+			// Successfully generated the preview.
+			try {
+				const blob = await response.blob();
+				resolve( window.URL.createObjectURL( blob ) );
+			} catch ( e ) {
+				resolve( mShotsEndpointUrl );
 			}
-		} )
-		.catch( reject ); // Pass through the reject notice
-};
+		} else {
+			reject();
+		}
+	} catch ( error ) {
+		reject( error ); // Pass through the reject notice.
+	}
+}
 
-export const loadmShotsPreview = ( options = {} ) => {
+export function loadmShotsPreview( options = {} ) {
 	return new Promise( ( resolve, reject ) => {
```

Comment:
don't need the wrapper Promise any more

Reply:
Most of it is quite doable, but then `querymShotsEndpoint` recursively calls itself with a `setTimeout` when retrying. It's possible to make that work by promisifying `setTimeout` and effectively building a promise chain with `await timeoutPromise; return await querymShotsEndpoint(...)` but I'm not sure that's more readable than passing the `resolve` and `reject` methods down.

We'd also need to add a catch block to `prefetchmShotsPreview` to avoid getting `Uncaught (in promise)` errors in the console (on Chrome, at least). I guess this is a useful debugging feature, but it makes throwing an error in an async function slightly different from calling reject on a promise :-/

Reply:
To perform some code in an async loop, Promise chains use (tail) recursion, while in async/await iteration is convenient. The `querymShots` async function can look like this:
```js
async function querymShots( options ) {
	const { maxRetries = 1, retryTimeout = 1000 } = options;

	for ( let retries = 0; ; retries++ ) {
		const result = await fetch();

		if ( isOk( result ) ) {
			return toBlobOrUrl( result );
		}

		if ( ! shouldRetry( result ) || retries++ >= maxRetries ) {
			throw new Error( `Still failing after ${ maxRetries } retries` );
		}

		// wait a bit before retrying
		await new Promise( r => setTimeout( r, retryTimeout ) );
	}
}
```

I think it would be a nice cleanup. But the PR works even without it and is shippable.

And yes, the `prefetch` function would need to swallow the exceptions:
```js
querymShots.catch( noop );
```

Reply:
That makes a lot of sense, of course, iteration! Didn't occur to me :)

I rewrote it according to your suggestion.